realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp

```diff
@@ -160,103 +158,93 @@ struct ChangeCallback {
 
 static void finalize_object(jlong ptr)
 {
-    TR_ENTER_PTR(ptr);
     delete reinterpret_cast<ObjectWrapper*>(ptr);
 }
 
-static inline size_t do_create_row(jlong shared_realm_ptr, jlong table_ptr)
+static inline Obj do_create_row_with_primary_key(JNIEnv* env, jlong shared_realm_ptr, jlong table_ref_ptr,
+                                                    jlong pk_column_key, jlong pk_value, jboolean is_pk_null)
 {
     auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
-    auto& table = *(reinterpret_cast<realm::Table*>(table_ptr));
-    shared_realm->verify_in_write();
-#if REALM_ENABLE_SYNC
-    return sync::create_object(shared_realm->read_group(), table);
-#else
-    return table.add_empty_row();
-#endif
-}
-
-static inline size_t do_create_row_with_primary_key(JNIEnv* env, jlong shared_realm_ptr, jlong table_ptr,
-                                                    jlong pk_column_ndx, jlong pk_value, jboolean is_pk_null)
-{
-    auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
-    auto& table = *(reinterpret_cast<realm::Table*>(table_ptr));
+    TableRef table = TBL_REF(table_ref_ptr);
+    ColKey col_key(pk_column_key);
     shared_realm->verify_in_write(); // throws
-    if (is_pk_null && !TBL_AND_COL_NULLABLE(env, &table, pk_column_ndx)) {
-        return realm::npos;
+    if (is_pk_null && !COL_NULLABLE(env, table, pk_column_key)) {
+        return Obj();
     }
 
     if (is_pk_null) {
-        if (table.find_first_null(pk_column_ndx) != npos) {
+
+        if (bool(table->find_first_null(col_key))) {
             THROW_JAVA_EXCEPTION(env, PK_CONSTRAINT_EXCEPTION_CLASS, format(PK_EXCEPTION_MSG_FORMAT, "'null'"));
         }
     }
     else {
-        if (table.find_first_int(pk_column_ndx, pk_value) != npos) {
+        if (bool(table->find_first_int(col_key, pk_value))) {
             THROW_JAVA_EXCEPTION(env, PK_CONSTRAINT_EXCEPTION_CLASS,
                                  format(PK_EXCEPTION_MSG_FORMAT, reinterpret_cast<long long>(pk_value)));
         }
     }
 
-    size_t row_ndx;
+    Obj obj;
 #if REALM_ENABLE_SYNC
     if (is_pk_null) {
-        row_ndx = sync::create_object_with_primary_key(shared_realm->read_group(), table, util::none);
+        obj = sync::create_object_with_primary_key(static_cast<Transaction&>(shared_realm->read_group()), *table, util::none);
```

Comment:
All this could be replaced by:
```
Mixed pk_val = is_pk_null ? Mixed() : Mixed(pk_value);
obj = table->create_object_with_primary_key(pk_val);
```